common/primitive/atomic_u128.rs

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright (C) 2023 Vaticle
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+use std::cell::Cell;
+use std::sync::atomic::AtomicU64;
+use std::sync::atomic::Ordering::SeqCst;
+
+/// Implements an AtomicU128 restricted to increments of u32.
+/// The lower-word (LW) is implemented using an AtomicU64. It is always correct.
+/// The upper word (UW) is implemented using u128, with an overlap of the upper 8 bits of the lower-word. The upper word + 8 sync-bits are eventually correct.
+/// The overlap is used to detect the UW being outdated.
+/// We restrict increments to at most u32::MAX, so that it is reasonable to assume:
+/// 1. there are never concurrent updates to the sync-bits or to UW.
+/// 2. The UW is out of sync by exactly 1 when sync_bits(LW) < sync_bits(UW), and is up-to-date otherwise
+///
+/// It is the responsibility of the thread that updates the sync_bits (and possibly overflow) to update sync_bits in the u128.
+/// Other threads can return (UW+1)(LW) if sync_bits(LW) < sync_bits(UW);  (UW)(LW) otherwise
+pub struct AtomicU128 {
+    uw_sync: Cell<u128>,
+    lw: AtomicU64,
+}
+
+const SYNC_MASK: u64 = 0xff00_0000_0000_0000;
+const UW_MASK: u128 = 0xffff_ffff_ffff_ffff__0000_0000_0000_0000;
+const UW_INCREMENT: u128 = 0x0000_0000_0000_0001__0000000000000000;
+
+impl AtomicU128 {
+
+    pub fn new(from: u128) -> Self {
+        AtomicU128 {
+            uw_sync: Cell::new(from),
+            lw: AtomicU64::new(from as u64)
+        }
+    }
+
+    #[inline]
+    fn uw_outdated(uw_sync: u128, lw: u64) -> bool {
+        (lw & SYNC_MASK) < ((uw_sync as u64) & SYNC_MASK)
+    }
+
+    #[inline]
+    fn compose(uw_sync: u128, lw: u64) -> u128 {
+        (uw_sync & UW_MASK) | (lw as u128)
+    }
+
+    pub fn load(&self) -> u128 {
+        let uw_read = self.uw_sync.get();
+        let lw_read  = self.lw.load(SeqCst);
+        let uw_increment = if Self::uw_outdated(uw_read, lw_read) { UW_INCREMENT } else { 0 };
+        Self::compose(uw_read + uw_increment, lw_read)
+    }
+
+    pub fn add_fetch(&self, increment: u32) -> u128 {
+        let uw_read = self.uw_sync.get();
```

Comment:
@dmitrii-ubskii, Is reading a u128 atomic? I think I assumed it would be, but I don't see why that would be the case. If so, I'll have to revise the whole thing to use a u64.